scheduler/scheduler.go

```diff
@@ -220,10 +220,139 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
+	}
+
+	// parents come in sorted order from soonest to latest expected finish time
+	for i := len(allParents) - 1; i >= 0; i-- {
+		parent := allParents[i]
+		currentContainers, err := parent.GetContainers()
+		if err != nil {
+			return host.Host{}, errors.Wrapf(err, "Could not find containers for parent %s", parent.Id)
+		}
+		if len(currentContainers) < d.MaxContainers {
+			return parent, nil
+		}
+	}
+	return host.Host{}, errors.New("No available parent found for container")
+}
+
+// numNewParentsNeeded returns the number of additional parents needed to
+// accommodate new containers
+func numNewParentsNeeded(numCurrentParents, numContainersNeeded, numExistingContainers int, d distro.Distro) int {
+	if numCurrentParents*d.MaxContainers < numExistingContainers+numContainersNeeded {
+		return int(math.Ceil(float64(numContainersNeeded) / float64(d.MaxContainers)))
+	}
+	return 0
+}
+
+// parentCapacity calculates number of new parents to create
+// checks to make sure we do not create more parents than allowed
+func parentCapacity(d distro.Distro, numNewParents, numCurrentParents, numCurrentContainers, numContainersToSpawn int) int {
+	// if there are already maximum numbers of parents running, only spawn
+	// enough containers to reach maximum capacity
+	if numCurrentParents >= d.PoolSize {
+		numNewParents = 0
+	}
+	// if adding all new parents results in more parents than allowed, only add
+	// enough parents to fill to capacity
+	if numNewParents+numCurrentParents > d.PoolSize {
+		numNewParents = d.MaxContainers - numCurrentParents
+	}
+	return numNewParents
+}
+
+// containerCapacity calculates how many containers to make
+// checks to make sure we do not create more containers than can fit currently
+func containerCapacity(d distro.Distro, numCurrentParents, numCurrentContainers, numContainersToSpawn int) int {
+	if numContainersToSpawn > 0 {
+		numAvailableContainers := numCurrentParents*d.MaxContainers - numCurrentContainers
+		if numContainersToSpawn > numAvailableContainers {
+			numContainersToSpawn = numAvailableContainers
+		}
+	}
+	return numContainersToSpawn
+}
+
+// insertParents creates host intent documents for each parent
+func insertParents(d distro.Distro, numNewParents int) ([]host.Host, error) {
+	hostsSpawned := make([]host.Host, 0)
+	for j := 0; j < numNewParents; j++ {
+		parentHostOptions := generateParentHostOptions()
+
+		intentHost := cloud.NewIntent(d, d.GenerateName(), d.Provider, parentHostOptions)
+		if err := intentHost.Insert(); err != nil {
+			grip.Error(message.WrapError(err, message.Fields{
+				"runner":   RunnerName,
+				"host":     intentHost.Id,
+				"provider": d.Provider,
+			}))
+			return nil, errors.Wrapf(err, "Could not insert parent '%s'", intentHost.Id)
+		}
+		hostsSpawned = append(hostsSpawned, *intentHost)
+	}
+	return hostsSpawned, nil
+}
+
+// insertContainer creates a host intent document for a container
+func insertContainer(d distro.Distro) (*host.Host, error) {
```

Comment:
This is used for inserting host documents for distros that don't use containers right? I would rename this to something like "insertIntent" and mention that it inserts both in the comments